### PR TITLE
debug: test and fix race conditions in debug endpoints

### PR DESCRIPTION
**Pull request by howardjohn**
This makes it so during a few tests we call all the debug endpoints.

This found a few races which have been fixed in this PR

### CODE REVIEW

File: pilot/pkg/xds/ads.go [hunk 1]
```diff
@@ -386,6 +386,8 @@ func (s *DiscoveryServer) initializeProxy(con *Connection) error {
 }
 
 func (s *DiscoveryServer) computeProxyState(proxy *model.Proxy, request *model.PushRequest) {
+	proxy.Lock()
+	defer proxy.Unlock()
```
Comment (hzxuzhonghu) [inline pilot/pkg/xds/ads.go:390]: What does it really protect? If it is to protect the whole proxy struct, should be called at more places

File: pkg/kube/krt/collection.go [hunk 2]
```diff
@@ -343,6 +343,8 @@ func (h *manyCollection[I, O]) onPrimaryInputEvent(items []Event[I]) {
 func (h *manyCollection[I, O]) onPrimaryInputEventLocked(items []Event[I]) {
 	var events []Event[O]
 	recomputedResults := make([]map[Key[O]]O, len(items))
+
+	pendingDepStateUpdates := make(map[Key[I]]*collectionDependencyTracker[I, O], len(items))
```
Comment (hzxuzhonghu) [inline pkg/kube/krt/collection.go:347]: the collection fix need a test

Comment (howardjohn) [reply pkg/kube/krt/collection.go:347]: it's tested by the test added in the PR - that is how I found it

File: pilot/pkg/xds/ads.go (hunk 1 above)
Comment (howardjohn) [reply pilot/pkg/xds/ads.go:390]: There are like 6 fields in this function that are set - any one of them races with the debug handler that tries to read them

Comment (hzxuzhonghu) [reply pilot/pkg/xds/ads.go:390]: There are still many `read` in debug.go that not protected, do they not need to protect?
```
			Labels:       c.proxy.Labels,
			Metadata:     c.proxy.Metadata,
			Locality:     c.proxy.Locality,
```

```
proxy.LastPushContext,
```

```
proxy.SidecarScope
```

File: pkg/kube/krt/collection.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pkg/kube/krt/collection.go:347]: The one in the workload_test.go?

Comment (howardjohn) [reply pkg/kube/krt/collection.go:347]: yes

File: pilot/pkg/xds/ads.go (hunk 1 above)
Comment (howardjohn) [reply pilot/pkg/xds/ads.go:390]: the debug code clones the proxy before accessing things so it's more racing during the clone than direct accesses to individual fields afaik

Comment (hzxuzhonghu) [reply pilot/pkg/xds/ads.go:390]: ah, right

**Review by hzxuzhonghu — APPROVED** [review]

**Pull request merged**